### ISSUE TITLE
Support ChainerX in all examples

### ISSUE THREAD

**Issue by niboshi** [chainer/chainer#7220]
We have been working on supporting new device specifiers in examples (#6661) and it's almost finished, but some examples do not support ChainerX yet, due to lack of ChainerX support in some Chainer features.

This is a tracking issue for those examples and features that need fixes to support ChainerX.

### Examples that do not support ChainerX yet
Note that for each example, only one feature that caused an error is listed. We may encounter another error after solving a single error.

* [x] `imagenet/train_imagenet_data_parallel.py` (a.)
* [x] `mnist/train_mnist_data_parallel.py` (b.) #7227
* [x] `mnist/train_mnist_model_parallel.py` (c.) #7330
* [x] ~`pix2pix/train_facade.py` (d.)~
* [x] ~`ptb/train_ptb.py` (d.)~
* [x] `ptb/gentext.py` (`ptb/train_ptb.py`) https://github.com/chainer/chainer/pull/7314
* [x] ~`ptb/train_ptb_custom_loop.py` (d.)~
* [x] `static_graph_optimizations/cifar/train_cifar.py` (h.) https://github.com/chainer/chainer/pull/7431
* [x] `static_graph_optimizations/cifar/train_cifar_custom_loop.py` (h.) https://github.com/chainer/chainer/pull/7431
* [x] `static_graph_optimizations/mnist/train_mnist.py` (h.) https://github.com/chainer/chainer/pull/7431
* [x] `static_graph_optimizations/mnist/train_mnist_custom_loop.py` (h.) https://github.com/chainer/chainer/pull/7431
* [x] ~`static_graph_optimizations/ptb/train_ptb_custom_loop.py` (h.)~

### Chainer features that need fixes for ChainerX

* [x] a. ChainerMN ~`MultiprocessParallellUpdater` does not support the new devices.~  #6000
  - This is out of scope of this issue.
* [x] b. `Variable.copydata()`  #7221
* [x] c. `StandardUpdater` transfers the model parameters, which causes device mismatch. #7324
   - Add ~`model_device` and~ `input_device` options to `StandardUpdater` ~(the existing `device` option will be an alias of the former)~, to specify the devices to transfer ~the model and~ the input data ~respectively~. If ~`model_device`~ `device` is omitted, it skips the model transfer. If `input_device` is omitted, ~`model_device`~ `device` is used to transfer the data.
* [ ] d. ~`Variable.unchain_backward()`~
   - `chainerx.backward()` unchains the graph unlike `chainer` counterpart. In `Variable.unchain_backward()`, if we can tell if the ChainerX array is disconnected, we can just do nothing. If it is connected, raise an error (the current behavior).
* [x] e. ~`BatchNormalization.forward` of the FunctionNode is called with chainerx.ndarrays.~
* [x] f. ~Accesses `Variable.node` in `StaticScheduleFunction.set_out_variables`~
* [x] g. ~Static graph with `EmbedID` does not work.~
* [x] h. Add an option to disable static graph optimization and use that option for ChainerX devices in the examples, with a warning. https://github.com/chainer/chainer/pull/7369 https://github.com/chainer/chainer/pull/7431

**Comment by hvy** [comment]
Notes about supporting examples relying on `unchain_backward` with ChainerX:
- No actions are required for now, keep those examples unsupported. Explanations follow.
- Unless `enable_double_backprop=True` is given to `chainerx.backward`, graphs are disconnected by default in ChainerX. Thus `unchain_backward` is practically not needed for the above TBPTT (truncated backprop through time) examples.
- However, this differs from Chainer, which preserves the graph. This difference in the behavior of backward is not clearly communicated to the user. We therefore would like to add an option to ChainerX's backward, not to disconnect the graph during the backward traversal. Chainer's backward should call the ChainerX backward with this flag properly set so that the graph remains connected. This will be a separate task.
- In the future, we might want to provide an "unchain_backward"-like function in ChainerX too, or, an option to Chainer's `backward` that disconnects the graph. In the latter cast, many TBPTT examples should use this flag with `backward` instead of relying on a seprate call to `unchain_backward`.

**Comment by hvy** [comment]
Filled in the check for 
> imagenet/train_imagenet_data_parallel.py (a.)

as it's intentionally not supported. We're exiting the example with a "not supported" message.

**Comment by hvy** [comment]
Closing as fixed.
